alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -187,8 +194,31 @@ checkCostModel cm =
 defaultCostModel :: Maybe CostModel
 defaultCostModel = CostModel <$> defaultCostModelParams
 
-instance FromCBOR CostModel where
-  fromCBOR = decode $ SumD checkCostModel <? (D mapFromCBOR)
+decodeCostModelMap :: Decoder s (Map Language CostModel)
+decodeCostModelMap = decodeMapByKey fromCBOR decodeCostModel
+
+decodeCostModel :: Language -> Decoder s CostModel
+decodeCostModel PlutusV1 =
+  case defaultCostModelParams of
+    Nothing -> fail "Default Plutus Cost Model is corrupt."
+    Just dcm -> do
+      checked <- checkCostModel <$> decodeArrayAsMap (Map.keysSet dcm) fromCBOR
+      case checked of
+        Left e -> fail e
+        Right cm -> pure cm
+
+decodeArrayAsMap :: Ord a => Set a -> Decoder s b -> Decoder s (Map a b)
```

Comment:
Okay, so we gave you the cost model as `Map Text Integer` to make it very homogenous, and so you could work with it somewhat independently of the API we give you. But if we're using `defaultCostModelParams` during deserialization, then that's kind of gone out the window. 

In which case: it seems like this would be easier if it was `Map ModelKey Integer` where `ModelKey` has a `Enum`/`Bounded` instances? It also would make for a nicer, more generic function:
```
decodeArrayAsMap :: (Enum k, Bounded k, Ord k) => Decoder s v -> Decoder s (Map k v)
```

Reply:
Yep, I agree, that would probably be nicer to work with and maintain.

Reply:
https://jira.iohk.io/browse/SCP-2696

Reply:
Thanks @michaelpj ! The enumeration will have to match what we are currently doing (ascending order by key) if we want to make this change sometime before the next hardfork after alonzo (since the cost model is a part of the script integrity hash).

Reply:
Good point!